src/app.py

```diff
@@ -1,62 +1,196 @@
-CONFUL_PATH = os.path.join(DIR, '../data', 'confluences.shp')
+REACH_PATH = os.path.join(DIR, "../data", "reaches.shp")
+BASIN_PATH = os.path.join(DIR, "../data", "basins.shp")
+CENTROID_PATH = os.path.join(DIR, "../data", "centroids.shp")
+CONFUL_PATH = os.path.join(DIR, "../data", "confluences.shp")
+
+# Load expected fields from JSON file
+with open(os.path.join(DIR, r"pyromb\resources", r"expected_fields.json"), "r") as f:
+    EXPECTED_FIELDS_JSON = json.load(f)
+
+# Convert JSON to the required dictionary format
+EXPECTED_FIELDS = {
+    key: [(field["name"], field["type"]) for field in fields]
+    for key, fields in EXPECTED_FIELDS_JSON.items()
+}
+
 
 class SFVectorLayer(sf.Reader, pyromb.VectorLayer):
     """
     Wrap the shapefile.Reader() with the necessary interface
-    to work with the builder. 
+    to work with the builder.
     """
+
     def __init__(self, path) -> None:
         super().__init__(path)
+        # Extract field names, skipping the first DeletionFlag field
+        self.field_names = [field[0] for field in self.fields[1:]]
+        # Precompute Shapely geometries for all shapes
+        self.shapely_geometries = [
+            shapely_shape(self.shape(i).__geo_interface__) for i in range(len(self))
+        ]
 
     def geometry(self, i) -> list:
         return self.shape(i).points
-    
+
+    def shapely_geometry(self, i):
+        """
+        Return the Shapely geometry for the ith shape.
+        """
+        return self.shapely_geometries[i]
+
     def record(self, i) -> dict:
-        return super().record(i)
-    
+        """
+        Return a dictionary mapping field names to their corresponding values.
+        """
+        rec = super().record(i)
+        return dict(zip(self.field_names, rec))
+
     def __len__(self) -> int:
         return super().__len__()
 
-def main():
-    ### Config ###
-    plot = False # Set True of you want the catchment to be plotted
-    model = pyromb.RORB() # Select your hydrology model, either pyromb.RORB() or pyromb.WBNM()
+
+def main(
+    reach_path: str | None = None,
+    basin_path: str | None = None,
+    centroid_path: str | None = None,
+    confluence_path: str | None = None,
+    output_name: str | None = None,
+    plot: bool = False,
+    model: Any | None = None,
+) -> None:
+    """
+    Main function to build and process catchment data.
+
+    Parameters
+    ----------
+    reach_path : str
+        Path to the reaches shapefile.
+    basin_path : str
+        Path to the basins shapefile.
+    centroid_path : str
+        Path to the centroids shapefile.
+    confluence_path : str
+        Path to the confluences shapefile.
+    output_name : str
+        Name of the output file.
+    plot : bool
+        Whether to plot the catchment.
+    model : pyromb.Model
+        The hydrology model to use.
+    """
+    # Set default paths if not provided
+    reach_path = reach_path or REACH_PATH
```

Comment:
Happy for the vector paths to be passed directly to the main function as this would reduce the need for global variables within main.